README.md
# FC-Cycle

[![Build status](https://ci.appveyor.com/api/projects/status/615gjbnthy6cmtvi?svg=true)](https://ci.appveyor.com/project/naderm/fc-cycle)

Utility for collecting fraction in a cycle on the Gilson FC 204 Fraction Collector.

## Installation

To download a completely packaged Windows executable, visit our [releases page](https://github.com/white-lab/fc-cycler/releases). This software has only been tested on Windows as it required altering several serial flow control settings that require platform-dependent code.

## Hardware

In order to connect your computer to the FC 204, you will need a USB to RS-232 cable (We used a [TRENDnet USB to Serial Converter](https://www.amazon.com/TRENDnet-Converter-Prolific-Chipset-TU-S9/dp/B0007T27H8/)). For development of this program, we also used a Gilson GSIOC cable to connect the male end of the RS-232 to the male GSIOC port on the instrument. However, one may be able to use a more readily available DB9 female to female cable to connect the two male ports together.

No GSIOC to RS-232 Converter box, such as the Gilson 606, was required in order to successfully connect to the FC 204.

## Usage

To run the cycler, connect the FC 204 to your computer using a USB to Serial cable, double-click the executable, and fill out the short control options:

```
>FC-Cycle.exe
Serial Port Name (optional):
GSIOC ID (default 61):
Max Tubes (default: 20):
Delay in Minutes (default: 5):
Total Time in Minutes (default: 85):
Time per tube in Seconds (default: 60):
```


ChangeLog.md
# Change Log

## 1.4.0

Features

  - Prioritize COM ports with "USB-to-Serial" in their description.

## 1.3.0

Features

  - Added separate 32-bit and 64-bit builds.

## 1.2.0

Features

  - Cycle tubes (1-6, 17-22) to reduce arm movement.

Bug fixes

  - Send T001 twice initially to avoid skipping tube 1.

## 1.1.0

Features

  - Log the index, total, and tube position for each position.

Bug fixes

  - Don't go to X000, Y000 if delay is 0.

## 1.0.0

Initial Release


tests/cycle_test.py
from unittest import TestCase

from fc_cycle import main


class ImportTest(TestCase):
    def test_total_tubes(self):
        self.assertEqual(
            main.get_total_tubes(85 * 60, 5 * 60, 60),
            80,
        )
        self.assertEqual(
            main.get_total_tubes(85 * 60, 0, 60),
            85,
        )

    def test_tube_pos(self):
        # Rack length * 2 tubes
        self.assertEqual(
            main.get_tube_pos(0, 22),
            1,
        )
        self.assertEqual(
            main.get_tube_pos(10, 22),
            11,
        )
        self.assertEqual(
            main.get_tube_pos(11, 22),
            12,
        )
        self.assertEqual(
            main.get_tube_pos(21, 22),
            22,
        )
        self.assertEqual(
            main.get_tube_pos(22, 22),
            1,
        )

        # 40 tubes
        self.assertEqual(
            main.get_tube_pos(0, 40),
            1,
        )
        self.assertEqual(
            main.get_tube_pos(10, 40),
            11,
        )
        self.assertEqual(
            main.get_tube_pos(11, 40),
            12,
        )
        self.assertEqual(
            main.get_tube_pos(21, 40),
            22,
        )
        self.assertEqual(
            main.get_tube_pos(22, 40),
            23,
        )
        self.assertEqual(
            main.get_tube_pos(32, 40),
            33,
        )
        self.assertEqual(
            main.get_tube_pos(33, 40),
            34,
        )
        self.assertEqual(
            main.get_tube_pos(39, 40),
            40,
        )
        self.assertEqual(
            main.get_tube_pos(40, 40),
            1,
        )

        # 12 tubes
        self.assertEqual(
            main.get_tube_pos(0, 12),
            1,
        )
        self.assertEqual(
            main.get_tube_pos(5, 12),
            6,
        )
        self.assertEqual(
            main.get_tube_pos(6, 12),
            17,
        )
        self.assertEqual(
            main.get_tube_pos(11, 12),
            22,
        )
        self.assertEqual(
            main.get_tube_pos(12, 12),
            1,
        )

        # 20 tubes
        self.assertEqual(
            main.get_tube_pos(0, 20),
            1,
        )
        self.assertEqual(
            main.get_tube_pos(9, 20),
            10,
        )
        self.assertEqual(
            main.get_tube_pos(10, 20),
            13,
        )
        self.assertEqual(
            main.get_tube_pos(19, 20),
            22,
        )
        self.assertEqual(
            main.get_tube_pos(20, 20),
            1,
        )

        # Odd number of tubes
        self.assertEqual(
            main.get_tube_pos(0, 5),
            1,
        )
        self.assertEqual(
            main.get_tube_pos(4, 5),
            22,
        )
        self.assertEqual(
            main.get_tube_pos(2, 5),
            3,
        )
        self.assertEqual(
            main.get_tube_pos(3, 5),
            21,
        )
        self.assertEqual(
            main.get_tube_pos(5, 5),
            1,
        )


tests/import_test.py
from unittest import TestCase


class ImportTest(TestCase):
    def test_imports(self):
        from fc_cycle import (
            main,
        )


fc_cycle/gsioc.py

import logging
import serial
import serial.tools.list_ports
import ctypes
from serial import win32


LOGGER = logging.getLogger("fc_cycle.gsioc")


class HandleSerial(serial.Serial):
    def __init__(self, *args, **kwargs):
        self.gsioc_id = kwargs.pop("gsioc_id", None) or 61
        super(HandleSerial, self).__init__(*args, **kwargs)

    def _reconfigure_port(self):
        super(HandleSerial, self)._reconfigure_port()

        comDCB = win32.DCB()
        win32.GetCommState(self._port_handle, ctypes.byref(comDCB))

        comDCB.fOutxCtsFlow = 0
        comDCB.fOutxDsrFlow = 0

        if not win32.SetCommState(self._port_handle, ctypes.byref(comDCB)):
            raise serial.SerialException(
                "Cannot configure port, something went wrong. "
                "Original message: {!r}".format(ctypes.WinError())
            )

        win32.PurgeComm(self._port_handle, win32.PURGE_RXCLEAR)

    def write_line(self, text):
        LOGGER.info("writing line: {}".format(repr(text)))
        run_command(self, text.encode(), gsioc_id=self.gsioc_id)


def run_command(port, message, gsioc_id=61):
    to_write = (128 + gsioc_id).to_bytes(1, byteorder='big')
    LOGGER.info("id: {}, message: {}".format(gsioc_id, message))
    LOGGER.debug("w: {}".format(to_write))
    port.write(to_write)
    LOGGER.debug("r: {}".format(port.read(1)))
    win32.PurgeComm(port._port_handle, win32.PURGE_RXCLEAR)

    if len(message) > 1:
        to_write = b"\x0a"
        LOGGER.debug("w: {}".format(to_write))
        port.write(b"\x0a")
        LOGGER.debug("r: {}".format(port.read(1)))
        win32.PurgeComm(port._port_handle, win32.PURGE_RXCLEAR)

    for letter in message:
        LOGGER.debug("w: {}".format(letter.to_bytes(1, byteorder='big')))
        port.write(letter.to_bytes(1, byteorder='big'))
        LOGGER.debug("r: {}".format(port.read(1)))
        win32.PurgeComm(port._port_handle, win32.PURGE_RXCLEAR)

    if len(message) > 1:
        to_write = b"\x0d"
        LOGGER.debug("w: {}".format(to_write))
        port.write(to_write)
        LOGGER.debug("r: {}".format(port.read(1)))
        win32.PurgeComm(port._port_handle, win32.PURGE_RXCLEAR)

    while True:
        rep = port.read(1)
        if not rep:
            break
        rep = rep[0]
        LOGGER.debug("r: {}".format(rep.to_bytes(1, byteorder="big")))
        win32.PurgeComm(port._port_handle, win32.PURGE_RXCLEAR)

        to_write = b"\x06"
        LOGGER.debug("w: {}".format(to_write))
        port.write(to_write)
        if rep > 127:
            LOGGER.debug("break")
            break


def _open_serial(path, gsioc_id=None):
    return HandleSerial(
        path,
        gsioc_id=gsioc_id,
        baudrate=9600,
        bytesize=serial.EIGHTBITS,
        parity=serial.PARITY_EVEN,
        stopbits=serial.STOPBITS_ONE,
        xonxoff=False,
        rtscts=True,
        dsrdtr=True,
        timeout=0.1,
        write_timeout=.1,
    )


def get_serial(serial_name=None, gsioc_id=None):
    if serial_name:
        return _open_serial(serial_name, gsioc_id=gsioc_id)

    excepts = []
    
    com_ports = serial.tools.list_ports.comports()
    com_ports = [
        i.device for i in sorted(com_ports, key=lambda x: ('USB-to-Serial' not in x.description, x.device))
    ]
    
    LOGGER.info('Testing COM ports: {}'.format(com_ports))

    for path in com_ports:
        try:
            port = _open_serial(path, gsioc_id=gsioc_id)
        except serial.SerialException as e:
            excepts.append(e)
            continue
        else:
            return port

    raise Exception("Unable to find any valid serial ports", excepts)


fc_cycle/main.py
from __future__ import absolute_import, division

import argparse
import logging
import math
import sys
import time

from fc_cycle import __version__, gsioc

LOGGER = logging.getLogger("fc_cycle.main")


def _parse_args(args):
    """
    Parses arguments from a argv format.

    Parameters
    ----------
    args : list of str

    Returns
    -------
    argparse.ArgumentParser
    """
    parser = argparse.ArgumentParser(
        prog="FC-Cycle",
        description="Utility for collecting fraction in a cycle on the Gilson "
        "FC 204 Fraction Collector.",
    )
    parser.add_argument(
        "-v", "--verbose",
        action="count",
        help="Increase verbosity of output.",
    )
    parser.add_argument(
        "-q", "--quiet",
        action="count",
        help="Decrease verbosity of output.",
    )
    parser.add_argument(
        '-V', '--version',
        action="version",
        version="%(prog)s {}".format(__version__),
    )
    return parser, parser.parse_args(args)


def _set_verbosity(args):
    verbosity = (args.verbose or 0) - (args.quiet or 0)

    if verbosity <= -2:
        level = logging.CRITICAL
    elif verbosity == -2:
        level = logging.ERROR
    elif verbosity == -1:
        level = logging.WARNING
    elif verbosity == 0:
        level = logging.INFO
    elif verbosity > 0:
        level = logging.DEBUG

    logger = logging.getLogger('fc_cycle')
    logger.setLevel(logging.DEBUG)
    formatter = logging.Formatter(
        '%(asctime)s - %(name)s - %(levelname)s - %(message)s'
    )

    ch = logging.StreamHandler(sys.stdout)
    ch.setLevel(level)
    ch.setFormatter(formatter)
    logger.addHandler(ch)

    fh = logging.FileHandler('fc_cycle.log')
    fh.setLevel(logging.DEBUG)
    fh.setFormatter(formatter)
    logger.addHandler(fh)

    LOGGER.debug(sys.argv)
    LOGGER.debug(args)


def get_total_tubes(total_time, delay, tube_time):
    return math.ceil((total_time - delay) / tube_time)


def get_tube_pos(index, max_tubes, rack_length=11):
    tube_pos = (index % max_tubes) + 1

    if max_tubes < rack_length * 2:
        if tube_pos > math.ceil(max_tubes / 2):
            tube_pos += rack_length * 2 - max_tubes

    return tube_pos


def main(args):
    _, args = _parse_args(args)
    _set_verbosity(args)

    LOGGER.debug("Running FC-Cycle Version {}".format(__version__))

    serial_name = input("Serial Port Name (optional): ")
    gsioc_id = input("GSIOC ID (default 61): ")
    max_tubes = int(input("Max Tubes (default: 20): ") or 20)
    delay = int(input("Delay in Seconds (default: 600): ") or 600)
    total_time = int(input("Total Time in Minutes (default: 85): ") or 85)
    tube_time = int(input("Time per tube in Seconds (default: 60): ") or 60)

    total_time *= 60

    with gsioc.get_serial(serial_name, gsioc_id=gsioc_id) as ser:
        try:
            # Beep and display Hello World
            ser.write_line("%")
            ser.write_line("G010")
            ser.write_line("W0Hello")
            ser.write_line("W01World")

            if delay > 0:
                ser.write_line("X000")
                ser.write_line("Y000")
                time.sleep(delay)

            # Cycle through each tube
            total_tubes = get_total_tubes(total_time, delay, tube_time)

            ser.write_line("T001")

            for i in range(total_tubes):
                tube_pos = get_tube_pos(i, max_tubes)

                LOGGER.info(
                    "Tube {} / {} (Position: {})"
                    .format(i + 1, total_tubes, tube_pos)
                )
                ser.write_line("T{:03d}".format(tube_pos))

                time.sleep(tube_time)
        except KeyboardInterrupt as err:
            LOGGER.info("Closing connection")
        finally:
            # Reset the machine
            ser.write_line("X000")
            ser.write_line("Y000")
            time.sleep(1)
            ser.write_line("X000")
            ser.write_line("Y000")


if __name__ == "__main__":
    try:
        main(sys.argv[1:])
    except:
        LOGGER.error("Execution error!", exc_info=True)
        input()
        sys.exit(1)
